specification/dartLangSpec.tex

```diff
@@ -19684,31 +19677,87 @@ \subsection{Type Aliases}
 (\ref{overview}).%
 }
 
-\commentary{%
-A type alias can be used as a type annotation,
-as a return type or parameter type in a function declaration,
-in a function type,
-as a type argument,
-in a type test,
-in a type cast,
-and in an \ON{} clause of a \TRY{} statement.
+\LMHash{}%
+We define the notion that a type alias declaration
+\NoIndex{expands to a type variable}
+inductively as follows:
+When $j \in 1 .. s$, we say that a type alias declaration of the form
+\code{\TYPEDEF{} $F$<\TypeParametersStd> = $X_j$;}
+\NoIndex{expands to a type variable at position} $j$.
+Similarly, if $H$ denotes a type alias declaration
+that declares $k$ type parameters and
+expands to a type variable at position $i \in 1 .. k$,
+we say that the type alias declaration
+
+\noindent
+\code{\TYPEDEF{} $G$<\TypeParametersStd> = $H$<\List{T}{1}{s}>;}
 
-Consider the case where the body of a given type alias $F$
-is a \synt{typeName} that denotes a non-generic class,
-or it is a parameterized type that starts with
-a \synt{typeName} that denotes a generic class,
-or one of these cases occur indirectly via another type alias.
-In this case $F$ or a parameterized type that starts with $F$,
-whichever is not an error,
-can also be used to name a constructor in an instance creation expression
-(\ref{instanceCreation}),
-and it can be used as a superclass, mixin, or superinterface
-(\ref{superclasses}, \ref{superinterfaces}, \ref{mixinApplication}).
-Moreover, $F$ or a parameterized type that starts with $F$,
-whichever is not an error,
-can be used to invoke static methods of the denoted class.%
+\noindent
+\NoIndex{expands to a type variable at position} $j$
+when $T_i$ is $X_j$ for some $j \in 1 .. s$.
```

Comment:
Maybe not *is* *T<sub>*i*</sub>*, but if *T<sub>*i*</sub>* *expands to* *X*_<sub>*j*</sub>.

Example:
```dart
typedef H<X> = X;
typedef G<Y> = H<H<Y>>;
```
Here the type argument to `H` *is not* the type variable `Y`, but it does expand to it.

(We either need to be very strict about the separation between syntactic form and its expansion to "normal form"/a real type, or we need to be deliberately loose about it. We've generally moved towards the former).

So, should we formally define that a *type term* can *expand* to a *type*? (I don't care if it's big-step/small-step operational semantics or denotational semantics).

A type term is like a type, except that it can also contain references to type aliases. We reduce the term to a type by iteratively expanding an alias reference into its definition. Because of the "no cycles" rule, this will end (not necessarily in polynomial space, or time depending on the reduction order).
Since there is no branching logic and no recursion, the reduction result is reduction order independent.

We can then talk about a term, and what it reduces/expands to. We'd then have:

> A type alias `T<X1...Xn> = R` expands to a type variable at position *i* if R expands to X*i*.

and we won't need to redefine the expansion rules in this predicate too.